File: 0054-spiral-matrix/0054-spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        // move l to r with u as ref  u++ 
        // move u to d with r as ref r--
        // move r to l with d as ref d--;
        // move d to u with l as ref l++;

        int m=matrix.size(),n=matrix[0].size();
        int l=0,r=n-1,u=0,d=m-1;
        vector<int> res;
        while(u<=d && l<=r){
            for(int i=l;i<=r;i++) res.push_back(matrix[u][i]);
            u++;
            // cout<<u<<" "<<d<<" "<<l<<" "<<r<<endl;
            if(u>d) break;
            for(int i=u;i<=d;i++) res.push_back(matrix[i][r]);
            r--;
            if(r<l) break;
            // cout<<u<<" "<<d<<" "<<l<<" "<<r<<endl;
            for(int i=r;i>=l;i--) res.push_back(matrix[d][i]);
            d--;
            // cout<<u<<" "<<d<<" "<<l<<" "<<r<<endl;
            if(d<u) break;
            for(int i=d;i>=u;i--) res.push_back(matrix[i][l]);
            l++;
            // cout<<u<<" "<<d<<" "<<l<<" "<<r<<endl;
            if(l>r) break;
        }

        return res;
    }
};

File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m,n=len(matrix),len(matrix[0])
        top,down=0,m-1 
        left,right=0,n-1 
        res = []
        while top<=down and left<=right: 
            for i in range(left,right+1): 
                res.append(matrix[top][i])
            top+=1 
            if(top>down): break

            for i in range(top,down+1): 
                res.append(matrix[i][right])
            
            right-=1
            if(left>right): break

            for i in range(right,left-1,-1):
                res.append(matrix[down][i])

            down-=1 
            if(top>down): break 
            for i in range(down,top-1,-1): 
                res.append(matrix[i][left])
            left+=1
    
        return res